File: tests/qt_types_standalone/rust/src/qpersistentmodelindex.rs
// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0

use cxx_qt_lib::{QModelIndex, QPersistentModelIndex};

#[cxx::bridge]
mod qpersistentmodelindex_cxx {
    unsafe extern "C++" {
        include!("cxx-qt-lib/qpersistentmodelindex.h");
        type QPersistentModelIndex = cxx_qt_lib::QPersistentModelIndex;
    }

    extern "Rust" {
        fn construct_qpersistentmodelindex() -> QPersistentModelIndex;
        fn read_qpersistentmodelindex(i: &QPersistentModelIndex, expected: i32) -> bool;
        fn clone_qpersistentmodelindex(i: &QPersistentModelIndex) -> QPersistentModelIndex;
    }
}

fn construct_qpersistentmodelindex() -> QPersistentModelIndex {
    QPersistentModelIndex::from(&QModelIndex::default())
}

fn read_qpersistentmodelindex(i: &QPersistentModelIndex, expected: i32) -> bool {
    i.is_valid() && i.row() == expected
}

fn clone_qpersistentmodelindex(i: &QPersistentModelIndex) -> QPersistentModelIndex {
    i.clone()
}

File: crates/cxx-qt-lib/src/core/qpersistentmodelindex.cpp
// clang-format off
// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// clang-format on
// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0
#include "cxx-qt-lib/qpersistentmodelindex.h"

#include "../assertion_utils.h"

// QPersistentModelIndex is a single pointer to a QPersistentModelIndexData
// https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/itemmodels/qabstractitemmodel.h?h=v5.15.6-lts-lgpl#n143
// https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/itemmodels/qabstractitemmodel.h?h=v6.2.4#n243
assert_alignment_and_size(QPersistentModelIndex,
                          alignof(::std::size_t),
                          sizeof(::std::size_t));

static_assert(
  !::std::is_trivially_copy_assignable<QPersistentModelIndex>::value);
static_assert(
  !::std::is_trivially_copy_constructible<QPersistentModelIndex>::value);

static_assert(!::std::is_trivially_destructible<QPersistentModelIndex>::value);

static_assert(QTypeInfo<QPersistentModelIndex>::isRelocatable);
